libraries/lib-music-information-retrieval/GetMeterUsingTatumQuantizationFit.cpp

```diff
@@ -0,0 +1,476 @@
+   std::transform(
+      peakIndices.begin(), peakIndices.end(), peakDistances.begin(),
+      [&](int peakIndex) {
+         const auto shiftedIndex = peakIndex - lag;
+         const auto closestDiv = std::round(shiftedIndex / odfSamplesPerDiv);
+         // Normalized distance between 0 and 1:
+         const auto distance = 2 *
+                               (shiftedIndex - closestDiv * odfSamplesPerDiv) /
+                               odfSamplesPerDiv;
+         return distance < 0 ? -distance : distance;
+      });
+   // Calculate the score as the sum of the distances weighted by
+   // the odf values:
+   const auto weightedAverage =
+      std::inner_product(
+         peakDistances.begin(), peakDistances.end(), peakValues.begin(), 0.,
+         [](double a, double b) { return a + b; },
+         [&](double distance, float peakValue) {
+            return distance * peakValue;
+         }) /
+      peakSum;
+   return weightedAverage;
+}
+
+// Look at the amplitude across all scores. If our ODF is that of a musical
+// rhythm, then the contrast between good and bad quantizations should make a
+// big difference.
+double GetExperimentContrastMeasure(
+   const std::unordered_map<int, MIR::OnsetQuantization>& quantizations,
+   double minError)
+{
+   const auto maxErrorIt = std::max_element(
+      quantizations.begin(), quantizations.end(),
+      [](const std::pair<int, OnsetQuantization>& a,
+         const std::pair<int, OnsetQuantization>& b) {
+         return a.second.error < b.second.error;
+      });
+   const auto amplitude = maxErrorIt->second.error - minError;
+   return amplitude;
+}
+
+struct QuantizationExperiment
+{
+   const OnsetQuantization bestFit;
+   const double score;
+};
+
+QuantizationExperiment RunQuantizationExperiment(
+   const std::vector<float>& odf, const std::vector<int>& peakIndices,
+   const std::vector<float>& peakValues,
+   const std::vector<int>& possibleNumTatums,
+   QuantizationFitDebugOutput* debugOutput)
+{
+   const auto quantizations = [&]() {
+      std::unordered_map<int, OnsetQuantization> quantizations;
+      std::transform(
+         possibleNumTatums.begin(), possibleNumTatums.end(),
+         std::inserter(quantizations, quantizations.end()), [&](int numTatums) {
+            const auto lag = GetOnsetLag(odf, numTatums);
+            const auto distance = GetQuantizationDistance(
+               peakIndices, peakValues, odf.size(), numTatums, lag);
+            return std::make_pair(
+               numTatums, OnsetQuantization { distance, lag, numTatums });
+         });
+      return quantizations;
+   }();
+
+   const auto bestFitIt = std::min_element(
+      quantizations.begin(), quantizations.end(),
+      [](const std::pair<int, OnsetQuantization>& a,
+         const std::pair<int, OnsetQuantization>& b) {
+         return a.second.error < b.second.error;
+      });
+
+   // Read the normalized distance, a value between 0 and 1, as a probability.
```

Comment:
Is the normalized distance really a "probability" (vertical axis if you plot the distribution) or is it the "random variable" (horizontal axis)?

Reply:
You're right, it's not a probability, `1 - distance` it's just a score. The probability of the file being a loop given that score is what we use in the benchmarking to determine the thresholds.